Development/nmos/constraints.cpp

```diff
@@ -0,0 +1,165 @@
+                else if (nmos::fields::constraint_minimum(constraint) > nmos::fields::constraint_minimum(subconstraint))
+                {
+                    return false;
+                }
+            }
+            if (constraint.has_field(nmos::fields::constraint_maximum) && subconstraint.has_field(nmos::fields::constraint_maximum))
+            {
+                if (constraint.at(U("maximum")).has_field(nmos::fields::numerator))
+                {
+                    if (nmos::parse_rational(constraint.at(U("maximum"))) < nmos::parse_rational(subconstraint.at(U("maximum"))))
+                    {
+                        return false;
+                    }
+                }
+                else if (nmos::fields::constraint_maximum(constraint) < nmos::fields::constraint_maximum(subconstraint))
+                {
+                    return false;
+                }
+            }
+            // subconstraint enum values should match constraint
+            if (subconstraint.has_field(nmos::fields::constraint_enum))
+            {
+                const auto& subconstraint_enum_values = nmos::fields::constraint_enum(subconstraint).as_array();
+                if (subconstraint_enum_values.end() == std::find_if(subconstraint_enum_values.begin(), subconstraint_enum_values.end(), [&constraint](const web::json::value& enum_value)
+                {
+                    return match_constraint(enum_value, constraint);
+                }))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        // Constraint Set B is a subset of Constraint Set A if all of Parameter Constraints of Constraint Set B, except for meta, are present in Constraint Set A and each Parameter Constraint of Constraint Set B, except for meta, is a subconstraint of the according Parameter Constraint of Constraint Set A.
+        // Constraint B is a subconstraint of a Constraint A if:
+
+        // 1. Constraint B has enum keyword when Constraint A has it and enum of Constraint B is a subset of enum of Constraint A
+        // 2. Constraint B has enum or minimum keyword when Constraint A has minimum keyword and allowed values for Constraint B are less than allowed values for Constraint A
+        // 3. Constraint B has enum or maximum keyword when Constraint A has maximum keyword and allowed values for Constraint B are greater than allowed values for Constraint A
+        bool is_constraint_subset(const web::json::value& constraint_set, const web::json::value& constraint_subset)
+        {
+            using web::json::value;
+
+            const auto& param_constraints_set = constraint_set.as_object();
+            const auto& param_constraints_subset = constraint_subset.as_object();
+
+            return param_constraints_subset.end() == std::find_if(param_constraints_subset.begin(), param_constraints_subset.end(), [&](const std::pair<utility::string_t, value>& subconstraint)
```

Comment:
Yes, if I've understood the intention I think the logic here is backwards.

Consider:

`constraint_set` A:
```json
{
  "urn:x-nmos:cap:param1": { "minimum": 42 }
}
```

`constraint_set` B1:
```json
{
  "urn:x-nmos:cap:param1": { "minimum": 57 }
  "urn:x-nmos:cap:param2": { "enum": ["foo"] }
}
```

B1 is a constraint subset of A because it only makes things more constrained (fewer `param1` values are allowed, and `param2` is also now constrained)

`constraint_set` B2:
```json
{
  "urn:x-nmos:cap:param1": { "minimum": 37 }
}
```

B2 is NOT a constraint subset of A because it is less constrained (more `param1` values are allowed)

`constraint_set` B3:
```json
{
}
```

B3 is NOT a constraint subset of A because it is is less constrained (`param1` is not constrained at all)

Reply:
Yes, you are completely right. Fixed the inverted logic and added your example as a unit test.